activemodel/lib/active_model/attribute_registration.rb

```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/class/subclasses"
+require "active_model/attribute_set"
+require "active_model/attribute/user_provided_default"
+
+module ActiveModel
+  module AttributeRegistration # :nodoc:
+    extend ActiveSupport::Concern
+
+    module ClassMethods # :nodoc:
+      def attribute(name, type = nil, default: (no_default = true), **options)
+        type = resolve_type_name(type, **options) if type.is_a?(Symbol)
+
+        pending = pending_attribute(name)
+        pending.type = type
+        pending.default = default unless no_default
+
+        reset_default_attributes
+      end
+
+      def _default_attributes # :nodoc:
+        @default_attributes ||= build_default_attributes
+      end
+
+      def attribute_types # :nodoc:
+        @attribute_types ||= _default_attributes.cast_types.tap do |hash|
+          hash.default = Type.default_value
+        end
+      end
+
+      private
+        class PendingAttribute # :nodoc:
+          attr_accessor :type, :default
+
+          def apply_to(attribute)
+            attribute = attribute.with_type(type || attribute.type)
+            attribute = attribute.with_user_default(default) if defined?(@default)
+            attribute
+          end
+        end
+
+        def pending_attribute(name)
+          @pending_attributes ||= {}
+          @pending_attributes[resolve_attribute_name(name)] ||= PendingAttribute.new
+        end
+
+        def apply_pending_attributes(attribute_set)
+          superclass.send(__method__, attribute_set) if superclass.respond_to?(__method__, true)
+
+          @pending_attributes&.each do |name, pending|
```

Comment:
It looks like this may need a `defined?` check so that it doesn't warn in ruby 2.7

`/rails/activemodel/lib/active_model/attribute_registration.rb:51: warning: instance variable @pending_attributes not initialized`

Reply:
Indeed!  It's not as pretty, but at least we'll be able to remove it whenever we drop support for Ruby 2.7.  :smiley:   Thanks!